torch_geometric/transforms/add_metapaths.py

```diff
@@ -192,3 +193,187 @@ def _get_edge_weight(self, data: HeteroData,
+            assert all([
+                j[-1] == path[i + 1][0] for i, j in enumerate(path[:-1])
+            ]), f"Invalid sequence of node types in '{path}'"
+
+        self.metapaths = metapaths
+        self.drop_orig_edges = drop_orig_edges
+        self.keep_same_node_type = keep_same_node_type
+        self.drop_unconnected_nodes = drop_unconnected_nodes
+        self.sample_ratio = sample_ratio
+        if isinstance(walks_per_node, int):
+            walks_per_node = [walks_per_node] * len(metapaths)
+        else:
+            assert len(walks_per_node) == len(metapaths)
+        self.walks_per_node = walks_per_node
+
+    def __call__(self, data: HeteroData) -> HeteroData:
+        edge_types = data.edge_types  # save original edge types
+        data.metapath_dict = {}
+
+        for j, metapath in enumerate(self.metapaths):
+            for edge_type in metapath:
+                assert data._to_canonical(
+                    edge_type) in edge_types, f"'{edge_type}' not present"
+
+            src_node = metapath[0][0]
+            num_nodes = data[src_node].num_nodes
+            num_starts = round(num_nodes * self.sample_ratio)
+            row = start = torch.randperm(num_nodes)[:num_starts].repeat(
+                self.walks_per_node[j])
+
+            for i, edge_type in enumerate(metapath):
+                edge_index = data[edge_type].edge_index
+                adj = SparseTensor(row=edge_index[0], col=edge_index[1],
+                                   sparse_sizes=data[edge_type].size())
+                col, mask = self.sample(adj, start)
+                row, col = row[mask], col[mask]
+                start = col
+
+            new_edge_type = (metapath[0][0], f'metapath_{j}', metapath[-1][-1])
+            data[new_edge_type].edge_index = coalesce(torch.vstack([row, col]))
+            data.metapath_dict[new_edge_type] = metapath
+
+        if self.drop_orig_edges:
+            for i in edge_types:
+                if self.keep_same_node_type and i[0] == i[-1]:
+                    continue
+                else:
+                    del data[i]
+
+        # remove nodes not connected by any edge type.
+        if self.drop_unconnected_nodes:
+            new_edge_types = data.edge_types
+            node_types = data.node_types
+            connected_nodes = set()
+            for i in new_edge_types:
+                connected_nodes.add(i[0])
+                connected_nodes.add(i[-1])
+            for node in node_types:
+                if node not in connected_nodes:
+                    del data[node]
+
+        return data
+
+    @staticmethod
+    def sample(src: SparseTensor, subset: Tensor,
+               num_neighbors: int = 1) -> Tensor:
+
+        rowcount = src.storage.rowcount()[subset]
+        mask = rowcount > 0
+        offset = torch.zeros_like(subset)
+        offset[mask] = src.storage.rowptr()[subset[mask]]
+
+        rand = torch.rand((rowcount.size(0), num_neighbors),
```

Comment:
A bit confused by this. `torch.rand` returns numbers between `0-1`, is this your intention?

Reply:
Correct. This is to sample from the node neighborhoods. I followed the way how the random walks are sampled in [MetaPath2Vec](https://github.com/pyg-team/pytorch_geometric/blob/80c6c782edb58a17cbabd54ae69d19cc1ba7785a/torch_geometric/nn/models/metapath2vec.py#L234-L251)